pgxn/neon/walsender_hooks.c

```diff
@@ -0,0 +1,121 @@
+}
+
+static int NeonWALPageRead(
+    XLogReaderState *xlogreader,
+    XLogRecPtr targetPagePtr,
+    int reqLen,
+    XLogRecPtr targetRecPtr,
+    char *readBuf)
+{
+    // Wait for flush pointer to advance past our request
+    XLogRecPtr flushptr = NeonWALReadWaitForWAL(targetPagePtr + reqLen);
+    int count;
+    if (flushptr < targetPagePtr + reqLen)
+	return -1;
+
+    // Read at most XLOG_BLCKSZ bytes
+    if (targetPagePtr + XLOG_BLCKSZ <= flushptr)
+	count = XLOG_BLCKSZ;
+    else
+	count = flushptr - targetPagePtr;
+
+    // Sometimes walsender requests non-monotonic sequences of WAL. If that's
+    // the case, we have to reset streaming from remote at the correct position.
+    if(targetPagePtr != NeonWALReaderGetRemLsn(wal_reader))
+    {
+	NeonWALReaderResetRemote(wal_reader);
+    }
+
+    static WaitEventSet *wait_event_set;
+    for(;;)
+    {
+	NeonWALReadResult res = NeonWALRead(
+	    wal_reader,
+	    readBuf,
+	    targetPagePtr,
+	    count,
+	    walprop_pg_get_timeline_id());
+	if(res == NEON_WALREAD_SUCCESS)
+	{
+            // Free event set, since we'll have to recreate it for a new socket later anyway
+	    if (NeonWALReaderEvents(wal_reader) == 0 && wait_event_set != NULL)
+	    {
+		FreeWaitEventSet(wait_event_set);
+		wait_event_set = NULL;
+	    }
+            // Required by xlogreader
+	    xlogreader->seg.ws_tli = NeonWALReaderGetSegment(wal_reader)->ws_tli;
+	    xlogreader->seg.ws_segno = NeonWALReaderGetSegment(wal_reader)->ws_segno;
+	    xlogreader->seg.ws_file = NeonWALReaderGetSegment(wal_reader)->ws_file;
+	    return count;
+	}
+	if(res == NEON_WALREAD_ERROR)
+	{
+	    return -1;
+	}
+        // Res is WOULDBLOCK, so we wait on the socket, recreating event set if necessary
+	{
+	    pgsocket sock = NeonWALReaderSocket(wal_reader);
+	    WaitEvent event;
+	    long timeout_ms = 1000;
+
+	    if (wait_event_set == NULL)
+	    {
+		wait_event_set = CreateWaitEventSet(TopMemoryContext, 3);
+		AddWaitEventToSet(wait_event_set, WL_LATCH_SET, PGINVALID_SOCKET, MyLatch, NULL);
+		AddWaitEventToSet(wait_event_set, WL_EXIT_ON_PM_DEATH, PGINVALID_SOCKET, NULL, NULL);
+		AddWaitEventToSet(wait_event_set, WL_SOCKET_WRITEABLE | WL_SOCKET_READABLE, sock, NULL, NULL);
+	    }
+	    WaitEventSetWait(wait_event_set, timeout_ms, &event, 1, WAIT_EVENT_WAL_SENDER_MAIN);
+	}
+    }
+}
+
+static void NeonWALReadSegmentOpen(XLogReaderState *xlogreader, XLogSegNo nextSegNo, TimeLineID *tli_p)
+{
+    neon_wal_segment_open(wal_reader, nextSegNo, tli_p);
+    xlogreader->seg.ws_file = NeonWALReaderGetSegment(wal_reader)->ws_file;
+}
+
+static void NeonWALReadSegmentClose(XLogReaderState *xlogreader)
+{
+    neon_wal_segment_close(wal_reader);
+    xlogreader->seg.ws_file = NeonWALReaderGetSegment(wal_reader)->ws_file;
+}
+
+void NeonOnDemandXLogReaderRoutines(XLogReaderRoutine *xlr)
+{
+    if(!wal_reader)
+    {
+        XLogRecPtr epochStartLsn = pg_atomic_read_u64(&GetWalpropShmemState()->propEpochStartLsn);
```

Comment:
Might be redundant, but better check that propEpochStartLsn is not 0.

Reply:
What should check do? Block until nonzero? Or throw error?

Reply:
ERROR is ok

Reply:
Added error